src/Symfony/Component/HttpKernel/Fragment/FragmentUriGenerator.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Fragment;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\UriSigner;
+
+/**
+ * Generates a fragment URI.
+ *
+ * @author Kévin Dunglas <kevin@dunglas.fr>
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+final class FragmentUriGenerator implements FragmentUriGeneratorInterface
+{
+    private $fragmentPath;
+    private $signer;
+
+    public function __construct(string $fragmentPath, UriSigner $signer = null)
+    {
+        $this->fragmentPath = $fragmentPath;
+        $this->signer = $signer;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function generate($controller, Request $request, bool $absolute = false, bool $strict = true, bool $sign = true): string
+    {
+        if ($sign && null === $this->signer) {
+            throw new \LogicException('You must use a URI when using the ESI rendering strategy or set a URL signer.');
+        }
+
+        if ($strict) {
+            $this->checkNonScalar($controller->attributes);
```

Comment:
this is broken if `$controller` is a string

Reply:
btw, the while method seems broken in this case

Reply:
Fixed in the meantime, this was a remaining of an experiment but it's useless.